pkg/services/cloudmigration/cloudmigrationimpl/xorm_store.go

```diff
@@ -252,6 +252,69 @@ func (ss *sqlStore) GetSnapshotList(ctx context.Context, query cloudmigration.Li
 	return snapshots, nil
 }
 
+func (ss *sqlStore) CreateUpdateSnapshotResources(ctx context.Context, snapshotUid string, resources []cloudmigration.CloudMigrationResource) error {
+	// ensure snapshot_uids are consistent so that we can use them to query when uid isn't known
+	for i := 0; i < len(resources); i++ {
+		resources[i].SnapshotUID = snapshotUid
+	}
+
+	return ss.db.InTransaction(ctx, func(ctx context.Context) error {
+		sql := "UPDATE cloud_migration_resource SET status=?, error_string=? WHERE uid=? OR (snapshot_uid=? AND resource_uid=?)"
+		err := ss.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+			for _, r := range resources {
+				// try an update first
+				result, err := sess.Exec(sql, r.Status, r.Error, r.UID, snapshotUid, r.RefID)
+				if err != nil {
+					return err
+				}
+				// if this had no effect, assign a uid and insert instead
+				n, err := result.RowsAffected()
+				if err != nil {
+					return err
+				} else if n == 0 {
+					r.UID = util.GenerateShortUID()
+					_, err := sess.Insert(r)
+					if err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("updating resources: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (ss *sqlStore) GetSnapshotResources(ctx context.Context, snapshotUid string, offset int, limit int) ([]cloudmigration.CloudMigrationResource, error) {
+	var resources []cloudmigration.CloudMigrationResource
+	if limit == 0 {
+		return resources, nil
+	}
+	err := ss.db.WithDbSession(ctx, func(sess *db.Session) error {
+		sess.Limit(limit, offset)
```

Comment:
People usually paginate by letting the client specify  something like this:

page=1 limit=10 // returns the first 10 items
page=2 limit=10 // returns items 11 to 20
page=3 limit=10 // returns items 11 to 30

Reply:
gotcha, I'll change this and the other spot where we do pagination